node/core/dispute-coordinator/src/real/participation/queues.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::{BTreeMap, HashMap};
+
+use thiserror::Error;
+
+use polkadot_primitives::v1::{CandidateHash, CandidateReceipt, SessionIndex};
+
+use crate::real::ordering::CandidateComparator;
+
+/// How many potential garbage disputes we want to queue, before starting to drop requests.
+const BEST_EFFORT_QUEUE_SIZE: usize = 100;
+
+/// How many priority disputes can be queued.
+///
+/// Once the queue exceeds that size, we will start to drop the newest participation requests in
+/// the queue. Note that for each vote import the request will be re-added, if there is free
+/// capacity. This limit just serves as a safe guard, it is not expected to ever really be reached.
+///
+/// For 100 parachains, this would allow for every single candidate in 100 blocks on
+/// two forks to get disputed, which should be plenty to deal with any realistic attack.
+const PRIORITY_QUEUE_SIZE: usize = 20_000;
+
+/// Type for counting how often a candidate was added to the best effort queue.
+type BestEffortCount = u32;
+
+/// Queues for dispute participation.
+pub struct Queues {
+	/// Set of best effort participation requests.
+	///
+	/// Note that as size is limited to `BEST_EFFORT_QUEUE_SIZE` we simply do a linear search for
+	/// the entry with the highest `added_count` to determine what dispute to participate next in.
+	///
+	/// This mechanism leads to an amplifying effect - the more validators already participated,
+	/// the more likely it becomes that more validators will participate soon, which should lead to
+	/// a quick resolution of disputes, even in the best effort queue.
+	best_effort: HashMap<CandidateHash, BestEffortEntry>,
+
+	/// Priority queue.
+	///
+	/// In the priority queue, we have a strict ordering of candidates and participation will
+	/// happen in that order.
+	priority: BTreeMap<CandidateComparator, ParticipationRequest>,
```

Comment:
why not use https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html?

Reply:
Mostly interface. `BinaryHeap` has a set like interface, while I need a map here, other than that `BinaryHeap` would be better here.

Reply:
But it is used as a set, because there is no lookup on `CandidateComparator`? So it's possible to create a wrapper
`OrderedParticipationRequest` which is a tuple with a comparison on the first field. And store it in the `BinaryHeap`.

Anyway, this is minor impl detail.

Reply:
Yeah, that's what I meant. Don't think the added code complexity is worth it here - disputes are not performance critical, they should just work if they happen.